iree/hal/detail.h

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_HAL_DETAIL_H_
+#define IREE_HAL_DETAIL_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "iree/base/api.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// Dispatches a method on a HAL object vtable.
+//
+// In the future we can use this to compile in a mode where all indirect
+// dispatches are replaced by direct calls to static methods. For example,
+// by changing the macro to resolve to `iree_hal_[resource]_[method_name]` we
+// can rely on LTO to perform cross-compilation unit inlining/strip unused HAL
+// calls/etc.
```

Comment:
Does this seemingly simple macro change work today? Can you elaborate a bit on the value and any tradeoffs here?

Reply:
Clarified TODO in the comment, but nope: still a few things to tweak - but I didn't want to do it until we needed it and could easily add a CI to keep it working. Pure WASM builds for web or embedded builds for particular hardware are examples of places that would benefit from this approach, while on the phone/desktop it's probably not as useful as being able to easily swap out backends/inject validation/profiling/etc.